### PR TITLE
narrow down gap for canister http_requests between ic-ref(-test) and Interface Spec

### PR DESCRIPTION
**Pull request by mraszyk**
This PR makes the following changes to ic-ref:
- add a flag `--cert-path` to import additional SSL certificates in `ic-ref`
- do not convert `RC_SYS_FATAL` and `RC_SYS_TRANSIENT` reject codes from management canister calls into `RC_CANISTER_ERROR` (these reject codes are produced by http_request calls)
- add EC_CANISTER_CONTRACT_VIOLATION error code
- update reject and error codes of management canister calls
- refine canister http_request method according to Interface Spec
- use custom `http-client` for making HTTP requests (the upstream client has too restricted limit on HTTP header size)

This PR makes the following changes to ic-ref-test:
- rename `poll_timeout` to `poll-timeout`
- add a flag `subnet-type` to vary certain constants depending on the subnet type (default: application)
- use a custom `httpbin` server for canister http_request tests
- do not expect protocol prefix (e.g., "https://") in `httpbin` flag
- fix invalid settings tests (use proper field names, use values just one above the maximum, restrict expected reject codes)
- add tests for new canister http_request checks in ic-ref

This PR makes the following changes to the universal-canister:
- add operation `Ops::TrapIfNeq` to trap with a given error message if two blobs are not equal
- add composite_query to sync universal_canister with ic repo

### CODE REVIEW

**Review by nomeata — APPROVED** [review]
Too long to review before breakfast, but description reads sensible

File: src/IC/Ref.hs [hunk 1]
```diff
@@ -930,29 +932,80 @@ invokeManagementCanister _ _ Closure{} = error "closure invoked on management ca
...
+  else if max_resp_size > max_response_bytes_limit then
+    reject RC_CANISTER_REJECT ("max_response_bytes cannot exceed " ++ show max_response_bytes_limit) (Just EC_CANISTER_REJECTED)
+  else if utf8_length (r .! #url) > max_http_request_url_length then
+    reject RC_CANISTER_REJECT "Failed to parse URL: uri too long" (Just EC_INVALID_ARGUMENT)
+  else if not (check_http_request_headers_number r) then
+    reject RC_CANISTER_REJECT ("number of request http headers exceeds the limit of " ++ show http_headers_max_number) (Just EC_CANISTER_REJECTED)
+  else if not (check_http_request_headers_name_length r) then
+    reject RC_CANISTER_REJECT ("number of bytes to represent some request http header name exceeds the limit of " ++ show http_headers_max_name_length) (Just EC_CANISTER_REJECTED)
+  else if not (check_http_request_headers_total_size r) then
+    reject RC_CANISTER_REJECT ("total number of bytes to represent request http headers exceeds the limit of " ++ show http_headers_max_total_size) (Just EC_CANISTER_REJECTED)
```
Comment (dsd-at-dfinity) [inline src/IC/Ref.hs:954]: nit: Is it possible to use guards to get rid of the repetitive `else if`?

File: src/IC/Test/Agent/Calls.hs [hunk 2]
```diff
@@ -150,33 +157,58 @@ ic_raw_rand :: HasAgentConfig => IC00 -> IO Blob
 ic_raw_rand ic00 =
   callIC ic00 agentEcid #raw_rand ()
 
-max_inter_canister_payload_in_bytes :: W.Word64
-max_inter_canister_payload_in_bytes = 2 * 1024 * 1024 -- 2 MiB
+ic_http_get_request ::
+    forall a b. (a -> IO b) ~ (ICManagement IO .! "http_request") =>
+    HasAgentConfig => ((W.Word64 -> W.Word64 -> W.Word64) -> IC00) -> String -> Maybe W.Word64 -> Maybe (String, Blob) -> Blob -> IO b
```
Comment (dsd-at-dfinity) [inline src/IC/Test/Agent/Calls.hs:162]: Could we have a descriptive type alias for this? `(W.Word64 -> W.Word64 -> W.Word64) -> IC00`? :-)

File: src/IC/Test/Options.hs [hunk 3]
```diff
@@ -28,31 +33,61 @@ instance IsOption ECID where
...
+newtype TestSubnetType = TestSubnetType SubnetType
 
-polltimeoutOption :: OptionDescription
-polltimeoutOption = Option (Proxy :: Proxy PollTimeout)
+instance Read TestSubnetType where
+  readsPrec _ x = do
+    if x == "application" then return (TestSubnetType Application, "")
+    else if x == "verified_application" then return (TestSubnetType VerifiedApplication, "")
+    else if x == "system" then return (TestSubnetType System, "")
+    else fail "could not read SubnetType"
```
Comment (dsd-at-dfinity) [inline src/IC/Test/Options.hs:75]: A perfect case for either guard or case expression ...

File: src/IC/Test/Agent/Calls.hs (hunk 2 above)
Comment (mraszyk) [reply src/IC/Test/Agent/Calls.hs]: Introduced `IC00WithCycles` and refactored.

File: src/IC/Ref.hs (hunk 1 above)
Comment (mraszyk) [reply src/IC/Ref.hs]: Using `MultiWayIf` now:
```
if
  | not transform_principal_check ->
    reject ...
  | max_resp_size > max_response_bytes_limit ->
    reject ...
```

**Review by dsd-at-dfinity — APPROVED** [review]
in the summary, you didn't mention the change regarding composite queries ... ;-) 

cool stuff!

**Comment by mraszyk** [comment]
> in the summary, you didn't mention the change regarding composite queries ... ;-)

Fixed!

**Pull request merged**